### PR TITLE
Synchronize CreateUser

### PR DESCRIPTION
**Pull request by kimlisa**
- Moved generateEmailToken function to email.go
- Unit tested generateEmailToken
- Unit tested race conditions for uuid/token generator
- golinted and formatted

### CODE REVIEW

File: service/service.go [hunk 1]
```diff
@@ -97,14 +110,20 @@ func (s *Service) CreateUser(ctx context.Context, req *pb.UserRequest) (*pb.User
...
-	// generate uuid
-	id, err := generateUUID()
+	// generate uuid synchronously to prevent users getting the same uuid
+	err := uuidGenerator.generateUUID()
 	if err != nil {
 		logger.Error("CreateUser generateUUID:", err.Error())
 		return nil, status.Error(codes.Unknown, err.Error())
 	}
-	user.Uuid = id
+	user.Uuid = uuidGenerator.uuid
```
[the remark this answers is not in the document]
Comment (faraonc) [reply service/service.go]: It is also unclear how uuid was derived from uuidGenerator.  
> Did we ‘generateUUID()?’

File: service/email_test.go [hunk 2]
```diff
@@ -176,3 +176,24 @@ func TestValidateEmail(t *testing.T) {
...
+		assert.Nil(t, err)
+		assert.NotEqual(t, "", tokenGenerator.token)
+
+		// test if key exists in the map
+		_, ok := tokens[tokenGenerator.token]
+		assert.Equal(t, false, ok)
+
+		tokens[tokenGenerator.token] = true
+	}
+
```
Comment (faraonc) [inline service/email_test.go]: You can do the following:
```
      // test each function call generats unique tokens
	tokens := make(map[string]bool)
	for i := 0; i < 10; i++ {
             go func {
		    err := tokenGenerator.generateEmailToken()
		    assert.Nil(t, err)
		    assert.NotEqual(t, "", tokenGenerator.token)

		    // test if key exists in the map
		    _, ok := tokens[tokenGenerator.token]
		    assert.Equal(t, false, ok)

		    tokens[tokenGenerator.token] = true
              }()
	}
```

Comment (faraonc) [reply service/email_test.go]: I am not totally sure, but this is close to the actual solution.

File: service/email_test.go [hunk 3]
```diff
@@ -176,3 +176,24 @@ func TestValidateEmail(t *testing.T) {
...
+	// test for race conditions
+	for i := 0; i < 10; i++ {
+		go tokenGenerator.generateEmailToken()
+	}
```
[the remark this answers is not in the document]
Comment (kimlisa) [reply service/email_test.go]: If i comment out the locks inside `generateEmailToken`, this test gives me race condition errors and fails the unit test

[the remark this answers is not in the document]
Comment (faraonc) [reply service/email_test.go]: Please see the latest comment, which should cover this block

File: service/email_test.go (hunk 2 above)
Comment (kimlisa) [reply service/email_test.go]: this works

File: service/service.go [hunk 4]
```diff
@@ -42,6 +45,9 @@ func init() {
 		currentServiceState: available,
 	}
 
+	//uuidGenerator = sync.Mutex{}
+	//tokenGenerator = sync.Mutex{}
```
Comment (faraonc) [inline service/service.go]: any reason why we have these?

**Pull request merged**